### PR TITLE
Define Fidelity spaces for benchmarks as ConfigurationSpace objects

### PR DESCRIPTION
**Pull request by NeoChaos12**
This is an initial implementation that essentially mimics the format used for defining the hyperparameter configuration spaces. It has been tested with the non-containerized XGBoost benchmark and tentative fidelity spaces have been added for non-containerized versions of all other benchmarks.

### CODE REVIEW

File: hpolib/abstract_benchmark.py [hunk 1]
```diff
@@ -114,7 +120,23 @@ def wrapper(self, configuration: Union[np.ndarray, ConfigSpace.Configuration, Di
                                 f'was {type(configuration)}')
 
             self.configuration_space.check_configuration(config)
-            return foo(self, configuration, **kwargs)
+
+            # Check the fidelities, if any
+            fidelity = self.opt_fidelity.get_dictionary()
+            if kwargs:
```
Comment (KEggensperger) [inline hpolib/abstract_benchmark.py]: You can rely on the fidelities being handed over as, e.g. fidelities={"epochs"=3}. So you can check for the existence of this dictionary in kwargs

File: hpolib/abstract_benchmark.py [hunk 2]
```diff
@@ -114,7 +120,23 @@ def wrapper(self, configuration: Union[np.ndarray, ConfigSpace.Configuration, Di
                                 f'was {type(configuration)}')
 
             self.configuration_space.check_configuration(config)
-            return foo(self, configuration, **kwargs)
+
+            # Check the fidelities, if any
+            fidelity = self.opt_fidelity.get_dictionary()
```
Comment (KEggensperger) [inline hpolib/abstract_benchmark.py]: Shouldn't this get get_fidelity_space?

File: hpolib/abstract_benchmark.py [hunk 3]
```diff
@@ -114,7 +120,23 @@ def wrapper(self, configuration: Union[np.ndarray, ConfigSpace.Configuration, Di
...
+            if kwargs:
+                # If kwargs contain any fidelity parameters, take them out
+                given_fidelity = {k: kwargs.pop(k, v) for k, v in fidelity.items()}
+
+                # Ensure that the extracted fidelity values play well with the defined fidelity space
```
Comment (KEggensperger) [inline hpolib/abstract_benchmark.py:152]: I would recommend the following:

1) Get the default configuration def from the fidelity space
2) Convert it to a dict
3) Convert the given fidelity to a dict, given_def or initialize as empty dict
4) Update def with values from given_def
5) Check whether it is a valid configuration given fidelity space

File: hpolib/benchmarks/nas/nasbench_101.py [hunk 4]
```diff
@@ -169,6 +169,30 @@ def _get_configuration_space(benchmark: Any, seed: Union[int, None] = None) -> C
...
+
+        Returns
+        -------
+        ConfigSpace.ConfigurationSpace
+        """
+        seed = seed if seed is not None else np.random.randint(1, 100000)
+        fidel_space = CS.ConfigurationSpace(seed=seed)
+
+        fidel_space.add_hyperparameters([
+            CS.CategoricalHyperparameter('budget', choices=[4, 12, 36, 108], default_value=108)
```
Comment (KEggensperger) [inline hpolib/benchmarks/nas/nasbench_101.py]: This should be a Ordinal

**Review by KEggensperger — CHANGES_REQUESTED** [review]
looks good, but we should change the objective interface to accept fidelities as dictionaries

File: hpolib/abstract_benchmark.py (hunk 3 above)
Comment (PhMueller) [reply hpolib/abstract_benchmark.py:152]: Also,I would suggest splitting it into a second wrapper function. So a single wrapper checking if the configuration is correct and one if the fidelity space is correct.

File: hpolib/abstract_benchmark.py (hunk 1 above)
Comment (PhMueller) [reply hpolib/abstract_benchmark.py]: Do you want to change the interface of the benchmarks? 
Currently, each objective function takes as input the fidelties as named parameters: 

e.g. objective_function(configuration, n_trees, subsample)

Changing it to objective_function(configuration, fidelities={'n_trees': 23, 'subsample': 0.5}) is in my opinion somehow less user friendly. 

To find the right parameters you than have to check first the fidelity space definition or the docstrings. 


Therefore, I like the approach in line 128.

File: hpolib/abstract_benchmark.py [hunk 5]
```diff
@@ -32,6 +32,11 @@ def __init__(self, rng: Union[int, np.random.RandomState, None] = None):
 
         self.rng = rng_helper.get_rng(rng=rng)
         self.configuration_space = self.get_configuration_space()
+        self.fidelity_space = self.get_fidelity_space()
+
+    @property
```
Comment (PhMueller) [inline hpolib/abstract_benchmark.py]: Why do you use @property here?
What is the use case for this function? 

Perhaps you could rename it to get_default_fidelities()

File: hpolib/abstract_benchmark.py (hunk 1 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: I agree with Phillip in that the current interface is more user friendly while the 'fidelities=...' approach would be more suitable for cases where a large number of fidelities are possible, thus creating potential for name clashes with other keyword arguments.

The change would also interfere with all existing code since it necessitates a change in the function call signature, although I'm not certain how much such code exists and how concerned we should be about interface changes.

File: hpolib/abstract_benchmark.py (hunk 2 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: Why should it? This might be wrapped up with the comments on line 128.

File: hpolib/abstract_benchmark.py (hunk 3 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py:152]: @KEggensperger If you check the definition of the opt_fidelity property, you will notice that the workflow here is essentially the same but more condensed. This workflow also takes into account the scenario where an objective function may be passed some keyword arguments (such as 'reset' from NASCifar10BaseBenchmark) that are not fidelity parameters. If no keyword arguments are given, it automatically sends the default fidelity forward, so the objective function can potentially be defined without optional fidelity arguments, although that's probably much less user-friendly.

@PhMueller I was wondering about that as well. I'll make the change.

File: hpolib/benchmarks/nas/nasbench_101.py (hunk 4 above)
Comment (NeoChaos12) [reply hpolib/benchmarks/nas/nasbench_101.py]: Aha, I missed that distinction. Will change it.

File: hpolib/abstract_benchmark.py (hunk 5 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: I think the default fidelity (or the fidelity to be optimized for in multi-fidelity opt) fits the role of a property more than a call to a one-line function. Of course, the difference between the two essentially boils down to saving the effort of typing two brackets when making a function call that takes no arguments and is also never expected to take any arguments.

Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: My original intention was to make it a class property, but that would involve some tinkering with the metaclass, so I opted to choose the simpler approach for now and shelve the idea for later.

Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: Defining it as a property in `AbstractBenchmark` itself has an additional advantage: it establishes a "default" relationship between the `default_configuration()` method of `get_configuration_space()` and the actual default fidelity used by any given benchmark, but which might be overwritten by any benchmarks, if they need to do so. I think I should rename it to `default_fidelity` though.

Comment (KEggensperger) [reply hpolib/abstract_benchmark.py]: We don't really use props in the rest of the code, so I would be in favour of making this consistent with `get_configuration_space()`, `get_fidelity_space()`

File: hpolib/abstract_benchmark.py (hunk 1 above)
Comment (KEggensperger) [reply hpolib/abstract_benchmark.py]: I thought it would be much easier for us (also in terms of testing) to rely on a unified interface which always takes the same arguments, `configuration, [fidelities, seed]`. I do understand that this will require to create a wrapper function for any optimizer we are currently aware of, but the interface is much cleaner. I need to think about this a bit more.

File: hpolib/abstract_benchmark.py (hunk 3 above)
Comment (KEggensperger) [reply hpolib/abstract_benchmark.py:152]: 1) Ah, I see, sorry I didn't look at this long enough. Could you rephrase the comment in l. 140 to say something along the lines of "for each fidelity use the given value in kwargs (and take them out) or its default value if it is not part of kwargs"
2) Yes, making this a separate function is definitely helpful

File: hpolib/abstract_benchmark.py (hunk 2 above)
Comment (KEggensperger) [reply hpolib/abstract_benchmark.py]: Is this still relevant? The comment is outdated?

File: hpolib/abstract_benchmark.py (hunk 1 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: I see the merits to the approach. Take your time, this change won't require a lot of effort from my end, at the very least. I'd just like to point out that we will need more changes to achieve total uniformity, as some of the benchmarks' objective functions have arguments that don't quite fit the bill for being fidelity parameters either. (e.g. NASCifar10BaseBenchmark's objective function has a boolean argument called 'reset').

My recommendation to deal with such settings is to have a separate benchmark-specific config dictionary/collection class with a more unified interface, so the objective function calls really only do exactly what they're supposed to do - accept a set of configurations to evaluate upon. But this is, again, hugely dependent upon whether or not it's worth the effort in our case, since such config collections are, by design, beneficial when dealing with a large number of settings.

Or perhaps the simplest of them all, move all such settings to class attributes of the specific Benchmarks themselves, and enforce that the objective function calls won't deal with any unknown keyword arguments.

File: hpolib/abstract_benchmark.py (hunk 2 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: Ah I missed that. Still new to the review interface. I'll resolve this conversation.

File: hpolib/abstract_benchmark.py (hunk 5 above)
Comment (NeoChaos12) [reply hpolib/abstract_benchmark.py]: Alright. Since this is a convenience method anyways, I would then just delete the property and default to using class.get_fidelity_space().get_default_configuration() whenever needed, as is the case with get_configuration_space().

**Comment by NeoChaos12** [comment]
Updates:

- The `fidelity: dict = None` based interface for objective function calls has been pushed through.
- All existing local benchmarks have been updated to use the new API.
- Updated API still needs testing.
- The container versions are expected to work with the new API automatically.

**Review by KEggensperger — CHANGES_REQUESTED** [review]
Could you please change the example, s.t. it makes use of 'fidelity' (and breaks)?
